sage/example-transition/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+//! Some example transitions.
+//!
+//! These should be expanded to really showcase the power of the SageApi design.
+
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use sage_api::{rules::ensure_asset_length, AssetT, SageApi, SageGameTransition};
+use scale_info::TypeInfo;
+use std::marker::PhantomData;
+
+/// Placeholder type, this was just a quick brain dump to get things going.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub struct Asset {
+	pub collection_id: u32,
+
+	pub asset_type: u32,
+
+	pub asset_sub_type: u32,
+
+	pub dna: [u8; 32],
+
+	pub minted_at: u32,
+}
+
+impl AssetT for Asset {
+	fn collection_id(&self) -> u32 {
+		self.collection_id
+	}
+
+	fn asset_type(&self) -> u32 {
+		self.asset_type
+	}
+
+	fn dna(&self) -> [u8; 32] {
+		self.dna
+	}
+
+	fn minted_at(&self) -> u32 {
+		self.minted_at
+	}
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub enum ExampleTransitionId {
+	UpgradeAsset,
+	ConsumeAsset,
+}
+
+pub struct ExampleTransition<AccountId, Balance> {
+	_phantom: PhantomData<(Balance, AccountId)>,
+}
+
+impl<AccountId, Balance> SageGameTransition for ExampleTransition<AccountId, Balance> {
+	type Asset = Asset;
+	type AccountId = AccountId;
+	type Balance = Balance;
+
+	type TransitionId = ExampleTransitionId;
+	type Extra = ();
+
+	fn verify_rule<
+		Sage: SageApi<Asset = Self::Asset, AccountId = Self::AccountId, Balance = Self::Balance>,
+	>(
```

Comment:
I am not really happy with this yet, but the problem is that it is hard in rust to combine generic code with non-generic code. So we will never get around constraining those associated types.

I am wondering if we should make the sage pallet less generic in general, but using the explicit runtime types that we are using in the ajuna runtime.

So the steps would be:
1. create a crate `ajuna_runtime_primitives`
2. make `SageApi` not generic with associated types, but us our runtime types here.

However, I am not 100% sure if this is a good idea. I will think some more.

Reply:
The reason why I am suggesting this is: our approach here, should make it as easy as possible for downstream developers to deploy a game, so ideally they should not care about generics too much, but currently, this is hard to avoid.

Reply:
But maybe I am overthinking and adding 

```rust
Sage: SageApi<Asset = Self::Asset, AccountId = Self::AccountId, Balance = Self::Balance>,
```
to every function that the game studio writes and needs, the `SageApi` is not too bad after all.